### ISSUE TITLE
Cannot read properties of null (reading 'redirect') when using Password Provider

### ISSUE THREAD

**Issue by xiaoqingming18** [get-convex/convex-auth#103]
I encountered an issue while developing the user registration feature using the **Password Provide**r from Convex Auth in a Next.js project. Upon submitting the registration form, the following error occurs:

`Cannot read properties of null (reading 'redirect')`

![PixPin_2024-10-19_20-19-50](https://github.com/user-attachments/assets/18c3130d-0e93-49d0-a706-b48dd3e0242a)

Additionally, the user registration data is not being submitted to the database.

Below is the relevant code from my project:

1. sign-up-card.tsx:
``` tsx
import React, { useState } from "react";
// Third-party tools/libraries
import { useAuthActions } from "@convex-dev/auth/react";
// Icons
import { TriangleAlert } from "lucide-react";
import { FcGoogle } from "react-icons/fc";
import { FaGithub } from "react-icons/fa";
// Components
import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "@/components/ui/card";
import { Input } from "@/components/ui/input";
import { Button } from "@/components/ui/button";
import { Separator } from "@/components/ui/separator";
// Interface/Types
import { SignInFlow } from "../types";

interface SignUpCardProps {
  setSignType: (signType: SignInFlow) => void;
}

export const SignUpCard = ({ setSignType }: SignUpCardProps) => {
  const { signIn } = useAuthActions();
  const [email, setEmail] = useState(""); // Email
  const [password, setPassword] = useState(""); // Password
  const [confirmPassword, setConfirmPassword] = useState(""); // Confirm Password
  const [pending, setPending] = useState(false); // Page loading state
  const [error, setError] = useState("");

  const handlePasswordSignUp = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (password !== confirmPassword) {
      setError("Passwords do not match");
      return;
    }

    setPending(true);
    signIn("password", { email, password, flow: "signup" }).catch((e) =>
      console.log(e)
    );
  };

  return (
    <Card className="w-full h-full p-8">
      <CardHeader className="px-0 pt-0">
        <CardTitle>Start by signing up for an account!</CardTitle>
      </CardHeader>
      <CardDescription className="mb-2">
        Sign up using your email or other methods
      </CardDescription>
      {!!error && (
        <div className="bg-destructive/15 p-3 rounded-md flex items-center gap-x-2 text-sm text-destructive mb-6">
          <TriangleAlert className="size-4" />
          <p>{error}</p>
        </div>
      )}
      <CardContent className="space-y-5 px-0 pb-0">
        <form className="space-y-2.5" onSubmit={handlePasswordSignUp}>
          <Input
            disabled={pending}
            value={email}
            onChange={(e) => setEmail(e.target.value)}
            placeholder="Email"
            type="email"
            required
          />
          <Input
            disabled={pending}
            value={password}
            onChange={(e) => setPassword(e.target.value)}
            placeholder="Password"
            // type="password"
            required
          />
          <Input
            disabled={pending}
            value={confirmPassword}
            onChange={(e) => setConfirmPassword(e.target.value)}
            placeholder="Confirm Password"
            // type="password"
            required
          />
          <Button type="submit" className="w-full" size="lg" disabled={pending}>
            Continue
          </Button>
        </form>
        <Separator />
        <div className="flex flex-col gap-y-2.5">
          <Button
            className="w-full relative"
            disabled={false}
            // onClick={() => handleProviderSignUp("google")}
            variant="outline"
            size="lg"
          >
            <FcGoogle className="size-5 absolute top-2.5 left-2.5" />
            Continue with Google
          </Button>
          <Button
            className="w-full relative"
            disabled={false}
            // onClick={() => handleProviderSignUp("github")}
            variant="outline"
            size="lg"
          >
            <FaGithub className="size-5 absolute top-2.5 left-2.5" />
            Continue with GitHub
          </Button>
        </div>
        <p className="text-xs text-muted-foreground">
          Already have an account?
          <span
            className="text-sky-700 hover:underline cursor-pointer"
            onClick={() => setSignType("signIn")}
          >
            Sign In
          </span>
        </p>
      </CardContent>
    </Card>
  );
};

```
2. Here is the content of my `package-lock.json`:
``` json
{
  "name": "team-link",
  "version": "0.1.0",
  "lockfileVersion": 3,
  "requires": true,
  "packages": {
    "": {
      "name": "team-link",
      "version": "0.1.0",
      "dependencies": {
        "@auth/core": "^0.36.0",
        "@convex-dev/auth": "^0.0.73",
        "@radix-ui/react-separator": "^1.1.0",
        "@radix-ui/react-slot": "^1.1.0",
        "class-variance-authority": "^0.7.0",
        "clsx": "^2.1.1",
        "convex": "^1.14.1",
        "lucide-react": "^0.453.0",
        "next": "14.2.15",
        "react": "^18",
        "react-dom": "^18",
        "react-icons": "^5.3.0",
        "tailwind-merge": "^2.5.4",
        "tailwindcss-animate": "^1.0.7"
      },
      "devDependencies": {
        "@types/node": "^20",
        "@types/react": "^18",
        "@types/react-dom": "^18",
        "eslint": "^8",
        "eslint-config-next": "14.2.15",
        "postcss": "^8",
        "tailwindcss": "^3.4.1",
        "typescript": "^5"
      }
    },
    ...
```

**Comment by syahirdev** [comment]
im facing the same issue. here is snippet of my SignInCard. When click github login button, the error is showing
`TypeError: Cannot read properties of null (reading 'redirect')`

```
export default function SignInCard({ setState }: Props) {
  const { signIn } = useAuthActions();

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  const onSignInWithProvider = (provider: 'github' | 'google') => {
    signIn(provider);
  };

return (
...
          <Button
            className="w-full relative"
            onClick={() => onSignInWithProvider('github')}
            variant="outline"
            size="lg"
            disabled={false}
          >
            <FaGithub className="size-7 absolute top-3 left-2.5" />
            Continue with GitHub
        </Button>
...
)
```

**Comment by kfrp** [comment]
Anyone solve this?

In deployment I am getting the same error with the convex auth `signIn` function from `useAuthActions`. Works fine in local, but on deployment throws `Uncaught (in promise) TypeError: Cannot read properties of null (reading 'redirect')`. 

Seems to be on [this line](https://github.com/get-convex/convex-auth/blob/c95f02700207a7679189ad3fbb1765512e085615/src/react/client.tsx#L203) in the convex code.

**Comment by sandervspl** [comment]
What fixes this error for me is adding `schema.ts` to the convex folder with the auth tables schema

```ts
import { authTables } from '@convex-dev/auth/server';
import { defineSchema, defineTable } from 'convex/server';
import { v } from 'convex/values';

const schema = defineSchema({
  ...authTables,
});

export default schema;
```

**Comment by Terris** [comment]
I have the same issue with Magic Links and NextJS App Router after upgrading from 0.0.74 to 0.0.80. But now I can't seem to find a downgrade that works. I surely must have something configured correctly because my Sign In with GitHub works well. 

## CLIENT CODE:

```
// /app/layout.tsx

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <ConvexAuthNextjsServerProvider>
      {/* `suppressHydrationWarning` only affects the html tag,
      and is needed by `ThemeProvider` which sets the theme
      class attribute on it */}
      <html lang="en" suppressHydrationWarning>
        <body className={`${inter.className} antialiased`}>
          <AppProviders>
            {children}
            <Toaster />
          </AppProviders>
        </body>
      </html>
    </ConvexAuthNextjsServerProvider>
  );
}

// AppProviders.tsx

"use client";

import { ConvexAuthNextjsProvider } from "@convex-dev/auth/nextjs";
import { ConvexReactClient } from "convex/react";
import { ThemeProvider } from "next-themes";
import { ReactNode } from "react";

const convex = new ConvexReactClient(process.env.NEXT_PUBLIC_CONVEX_URL!);

export function AppProviders({ children }: { children: ReactNode }) {
  return (
    <ThemeProvider
      attribute="class"
      defaultTheme="system"
      enableSystem
      disableTransitionOnChange
    >
      <ConvexAuthNextjsProvider client={convex}>
        {children}
      </ConvexAuthNextjsProvider>
    </ThemeProvider>
  );
}

// Magic Link Sign In Form in /app/page.tsx

import { useAuthActions } from "@convex-dev/auth/react";

function SignInWithMagicLink({
  handleLinkSent,
}: {
  handleLinkSent: () => void;
}) {
  const { signIn } = useAuthActions();
  const { toast } = useToast();
  return (
    <form
      className="flex flex-col"
      onSubmit={(event) => {
        event.preventDefault();
        const formData = new FormData(event.currentTarget);
        formData.set("redirectTo", "/dashboard");
        signIn("resend", formData)
          .then(handleLinkSent)
          .catch((error) => {
            console.error(error);
            toast({
              title: "Could not send sign-in link",
              variant: "destructive",
            });
          });
      }}
    >
      <label htmlFor="email">Email</label>
      <Input name="email" id="email" className="mb-4" autoComplete="email" />
      <Button type="submit">Send sign-in link</Button>
      <Toaster />
    </form>
  );
}
```

## CONVEX:

```
// auth.config.ts

const authConfig = {
  providers: [
    {
      domain: process.env.CONVEX_SITE_URL,
      applicationID: "convex",
    },
  ],
};
export default authConfig;


// auth.ts

import GitHub from "@auth/core/providers/github";
import Resend from "@auth/core/providers/resend";
import { convexAuth } from "@convex-dev/auth/server";

export const { auth, signIn, signOut, store, isAuthenticated } = convexAuth({
  providers: [GitHub, Resend],
});
```

## DEPS

```
// next app package.json:

{
  "name": "web",
  "version": "0.1.0",
  "type": "module",
  "private": true,
  "scripts": {
    "dev": "next dev --turbopack",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@auth/core": "^0.37.4",
    "@aws-sdk/client-s3": "^3.688.0",
    "@aws-sdk/credential-providers": "^3.687.0",
    "@convex-dev/auth": "^0.0.80",
    "@dnd-kit/core": "^6.1.0",
    "@dnd-kit/modifiers": "^9.0.0",
    "@dnd-kit/sortable": "^8.0.0",
    "@dnd-kit/utilities": "^3.2.2",
    "@radix-ui/react-dialog": "^1.1.2",
    "@radix-ui/react-slot": "^1.1.1",
    "@repo/convex": "workspace:*",
    "@repo/ui": "workspace:*",
    "@tanstack/react-table": "^8.20.6",
    "class-variance-authority": "^0.7.0",
    "clsx": "^2.1.1",
    "convex": "^1.18.2",
    "date-fns": "^4.1.0",
    "formik": "^2.4.6",
    "lucide-react": "^0.456.0",
    "next": "15.1.0",
    "next-themes": "^0.4.3",
    "react": "^19.0.0",
    "react-dom": "^19.0.0",
    "react-markdown": "^9.0.3",
    "tailwind-merge": "^2.5.4",
    "tailwindcss-animate": "^1.0.7",
    "yup": "^1.4.0",
    "zod": "^3.23.8"
  },
  "devDependencies": {
    "@repo/eslint-config": "workspace:^",
    "@repo/typescript-config": "workspace:*",
    "@tailwindcss/typography": "^0.5.16",
    "@types/node": "^20",
    "@types/react": "^18.3.0",
    "@types/react-dom": "^18.3.1",
    "postcss": "^8",
    "tailwindcss": "^3.4.1",
    "typescript": "^5"
  }
}


// convex package.json

{
  "name": "@repo/convex",
  "version": "0.0.0",
  "private": true,
  "main": "index.ts",
  "scripts": {
    "dev": "convex dev",
    "setup": "convex dev --until-success",
    "predev": "convex dev --until-success && node setup.mjs --once && convex dashboard",
    "build": "npx convex deploy"
  },
  "dependencies": {
    "@anthropic-ai/sdk": "^0.36.2",
    "@auth/core": "^0.37.4",
    "@aws-sdk/client-s3": "^3.734.0",
    "@aws-sdk/credential-providers": "^3.734.0",
    "@aws-sdk/lib-storage": "^3.734.0",
    "@convex-dev/auth": "^0.0.80",
    "@convex-dev/migrations": "^0.2.3",
    "@repo/eslint-config": "workspace:*",
    "@repo/typescript-config": "workspace:*",
    "convex": "^1.18.2",
    "convex-helpers": "~0.1.67",
    "resend": "^4.1.1",
    "sharp": "^0.33.5"
  },
  "devDependencies": {
    "@types/node": "^22.10.10",
    "@types/react": "^19.0.8",
    "@types/react-dom": "^19.0.3",
    "dotenv": "^16.4.7",
    "eslint": "^9.18.0",
    "typescript": "^5.7.3"
  }
}

```

**Comment by thomasballinger** [comment]
@Terris a lot has changed since this issue was opened, could you share the error you're seeing?

**Comment by Terris** [comment]
This is resolved.

**Comment by gitmibrahim** [comment]
@Terris How had it been resolved? I still face this with password provider

**Comment by MohamedKamran** [comment]
It's not fixed yet! Facing the same issue on Deployment it works well locally

**Comment by gitmibrahim** [comment]
I'm keep facing it locally!

**Comment by Gptimus** [comment]
> I'm keep facing it locally!

The same here !

**Comment by flowirtz** [comment]
Just leaving this here, because it helped me: I got the same exact error message on login, for my production deployment.
After a while I noticed an error in the Convex dashboard, and [then stumbled upon this guide](https://labs.convex.dev/auth/production). Turns out there are separate Convex Auth steps you need to follow to deploy the app.

TL;DR [Follow this guide if you haven't already](https://labs.convex.dev/auth/production)

**Comment by amoscicki** [comment]
I am stumbling across this error too. Not necessarily in the production, and not limited to password provider. I see this also with OTP for example. This is essentially returned when the authentication fails instead of getting the convex error related to failure reason. Auth works but error channel doesn't. 


> Seems to be on [this line](https://github.com/get-convex/convex-auth/blob/c95f02700207a7679189ad3fbb1765512e085615/src/react/client.tsx#L203) in the convex code.

Haven't had time to test this yet but have a feeling that adding a '?' before accessing the prop could fix that.

**Comment by opswiz** [comment]
This happened for me when I tried to signin with a email with invalid credentials on the password provider. 

I don't think this is something on implementation. Seems the response from signin is not handled properly.  The library should handle the result being null and throw proper ConvexError. 

Happens on a freshly created project with `npm create convex` with password auth added.

**Comment by amoscicki** [comment]
I submitted pull request with minimal fix, that uncovered that errors seem not to be passed. I have added a simplified passthrough as well, not sure if that will be deemed desired way by maintainers.
